### PR TITLE
Use file filters action instead of Github's files filter

### PR DESCRIPTION
**Pull request by lunny**
Inspired by https://github.com/go-gitea/gitea/pull/24530#issuecomment-1558815301

This PR use a file filter action to do different CI jobs according changed files types. All types are defined in `.github/file-filters.yml`. Now there are 4 types, `docs`, `backend`, `frontend` and `build`. Then if a PR only changed docs files, those CI jobs which passed the conditions will run, and other types are also like this.

### CODE REVIEW

File: .github/file-filters.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# This is used by the action https://github.com/dorny/paths-filter (which we have forked to https://github.com/dorny/paths-filter)
```
Comment (silverwind) [inline .github/file-filters.yml]: Fork URL wrong?

File: .github/workflows/pull-compliance-docs.yml [hunk 2]
```diff
@@ -2,16 +2,30 @@ name: compliance-docs
...
+    # Map a step output to a job output
+    outputs:
+      docs: ${{ steps.changes.outputs.docs }}
+    steps:
+      - uses: actions/checkout@v3  # v2
+      - name: Check for backend file changes
+        uses: dorny/paths-filter@v2  # v2
+        id: changes
+        with:
+          filters: .github/file-filters.yml
```
Comment (silverwind) [inline .github/workflows/pull-compliance-docs.yml]: Could even extract the whole job to a file.

https://docs.github.com/en/actions/using-workflows/reusing-workflows#example-reusable-workflow

File: .github/workflows/pull-compliance-docs.yml [hunk 3]
```diff
@@ -2,16 +2,30 @@ name: compliance-docs
...
 jobs:
+  files-changed:
+    name: detect what files changed
+    runs-on: ubuntu-latest
+    timeout-minutes: 3
+    # Map a step output to a job output
+    outputs:
+      docs: ${{ steps.changes.outputs.docs }}
+    steps:
+      - uses: actions/checkout@v3  # v2
```
Comment (silverwind) [inline .github/workflows/pull-compliance-docs.yml]: What are these `v2` comments signifying?

File: .github/file-filters.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+backend: &backend
+  - "**/*.go"
+  - "go.mod"
+  - "go.sum"
+
+frontend: &frontend
+  - "web_src/**"
+  - "package.json"
+  - "package-lock.json"
+  - "*.js"
```
Comment (silverwind) [inline .github/file-filters.yml]:
```suggestion
  - "**/*.js"
  - "web_src/**"
  - "package.json"
  - "package-lock.json"
```

File: .github/file-filters.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# This is used by the action https://github.com/dorny/paths-filter (which we have forked to https://github.com/dorny/paths-filter)
+
+docs: &docs
+  - "docs/**"
+  - "**/*.md"
```
Comment (silverwind) [inline .github/file-filters.yml]:
```suggestion
  - "**/*.md"
  - "docs/**"
```

File: .github/file-filters.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+backend: &backend
+  - "**/*.go"
```
Comment (silverwind) [inline .github/file-filters.yml:6]:
```suggestion
  - "**/*.go"
  - "**/*.tmpl"
```

**Comment by lunny** [comment]
It's ready to review now.

File: Makefile [hunk 7]
```diff
@@ -411,6 +413,10 @@ lint-go-vet:
 lint-editorconfig:
 	$(GO) run $(EDITORCONFIG_CHECKER_PACKAGE) templates .github/workflows
 
+.PHONY: lint-action
+lint-action:
+	actionlint
```
Comment (silverwind) [inline Makefile]:
```suggestion
	$(GO) run $(ACTIONLINT_PACKAGE)
```

File: Makefile [hunk 8]
```diff
@@ -199,6 +200,7 @@ help:
 	@echo " - deps-tools                       install tool dependencies"
 	@echo " - lint                             lint everything"
 	@echo " - lint-fix                         lint everything and fix issues"
+	@echo " - lint-action                      lint action workflow files"
```
Comment (silverwind) [inline Makefile]:
```suggestion
	@echo " - lint-actions                     lint action workflow files"
```

File: Makefile [hunk 9]
```diff
@@ -411,6 +413,10 @@ lint-go-vet:
 lint-editorconfig:
 	$(GO) run $(EDITORCONFIG_CHECKER_PACKAGE) templates .github/workflows
 
+.PHONY: lint-action
+lint-action:
```
Comment (silverwind) [inline Makefile]:
```suggestion
.PHONY: lint-actions
lint-actions:
```

File: Makefile [hunk 10]
```diff
@@ -35,6 +35,7 @@ SWAGGER_PACKAGE ?= github.com/go-swagger/go-swagger/cmd/swagger@v0.30.4
 XGO_PACKAGE ?= src.techknowlogick.com/xgo@latest
 GO_LICENSES_PACKAGE ?= github.com/google/go-licenses@v1.6.0
 GOVULNCHECK_PACKAGE ?= golang.org/x/vuln/cmd/govulncheck@latest
+ACTION_LINT_PACKAGE ?= github.com/rhysd/actionlint/cmd/actionlint@latest
```
Comment (silverwind) [inline Makefile]:
```suggestion
ACTIONLINT_PACKAGE ?= github.com/rhysd/actionlint/cmd/actionlint@latest
```

File: Makefile [hunk 11]
```diff
@@ -908,6 +914,7 @@ deps-tools:
 	$(GO) install $(XGO_PACKAGE)
 	$(GO) install $(GO_LICENSES_PACKAGE)
 	$(GO) install $(GOVULNCHECK_PACKAGE)
+	$(GO) install $(ACTION_LINT_PACKAGE)
```
Comment (silverwind) [inline Makefile]:
```suggestion
	$(GO) install $(ACTIONLINT_PACKAGE)
```

File: .github/workflows/files-changed.yml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    outputs:
+      docs:
+        description: "wehther docs files changed"
+        value: ${{ jobs.files-changed.outputs.docs }}
+      backend:
+        description: "wehther backend files changed"
+        value: ${{ jobs.files-changed.outputs.backend }}
+      frontend:
+        description: "wehther frontend files changed"
+        value: ${{ jobs.files-changed.outputs.frontend }}
```
Comment (silverwind) [inline .github/workflows/files-changed.yml]:
```suggestion
        description: "whether docs files changed"
        value: ${{ jobs.files-changed.outputs.docs }}
      backend:
        description: "whether backend files changed"
        value: ${{ jobs.files-changed.outputs.backend }}
      frontend:
        description: "whether frontend files changed"
        value: ${{ jobs.files-changed.outputs.frontend }}
```

File: .github/workflows/files-changed.yml [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+      backend:
+        description: "whether backend files changed"
+        value: ${{ jobs.files-changed.outputs.backend }}
+      frontend:
+        description: "whether frontend files changed"
+        value: ${{ jobs.files-changed.outputs.frontend }}
+
+jobs:
+  files-changed:
+    name: detect what files changed
```
Comment (silverwind) [inline .github/workflows/files-changed.yml]:
```suggestion
    name: detect which files changed
```

**Review by silverwind — APPROVED** [review]
Let's try it. For actionlint, we could add another filter category, btw.

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**